### PR TITLE
Bug 1864071 - Add debug drawer fab to an app-wide Composable overlay

### PR DESCRIPTION
**Pull request by MozillaNoah**
Note: this PR is based off of https://github.com/mozilla-mobile/firefox-android/pull/4520, so when reviewing this before the root PR lands, make sure to look at the commit just for 1864071 to keep to the actual changes of four files.

https://github.com/mozilla-mobile/firefox-android/assets/87384386/629334a3-7314-4a1a-9a96-7591b59ed686

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1864071

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:59]: For now, I think it'll make sense to include just the debug fab in this wrapper composable until we have more debugging UI to overlay the app. In other words, when the drawer UI is added, this might move into an intermediary wrapper `DebugOverlay`, if it makes sense to do so.

File: fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:5]: with this package addition, the thought is to have something like:

`debugdrawer/ui` for all of the Compose code
`debugdrawer/redux` for the store/state/middlewares
`debugdrawer/feature_name/` for any of the functionality (tab manager, secret settings, crash viewer, etc.)

Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:5]: perhaps we should use debugsettings as the overall feature name? Then Debug Drawer would be part of the UI package, wdyt?

File: fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:45]: we could use a "bug" type icon? 🙂

File: fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:51]: nit: Not sure about the positioning of the button, but appreciate this may change or is desirable in this use case

**Review by t-p-white — COMMENTED** [review]
Nice job, I'll hold off the final review until https://github.com/mozilla-mobile/firefox-android/pull/4520#event-11074801083 is merged

File: fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:45]: I'm actually working with Crystal to create an icon just for the drawer! [Ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1866128)

File: fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:51]: For 0.1, I'm thinking center of the screen. A follow-up once the drawer is stable is to make the FAB draggable, so it can be repositioned anywhere on the screen!

Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:51]: Maybe center-left to start? Think that would avoid the more common gesture-based back nav, and may leave more of the screen generally interact-able

File: fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:5]: Both of these make sense to me

File: fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugdrawer/ui/DebugDrawerFAB.kt:35]: nit: [typical java convention for acronyms of 3+ letters](https://stackoverflow.com/questions/2236807/java-naming-convention-with-acronyms) is for just the first letter to be uppercase. I see other fabs in the codebase follow that as well

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:287]: It might run a lower risk of performance impact to observe the data store here and call this whole stack conditionally

**Review by MatthewTighe — COMMENTED** [review]
Had a couple comments but this is looking pretty ready to me once the previous lands.

I do think we should [monitor for performance regressions](https://mozilla-hub.atlassian.net/wiki/spaces/Mobile/pages/222234014/Android+Performance+Regression+Monitoring) once this lands just so we can be aware of any impact. If there is any, we could consider shipping it only on debug builds first or something.

File: fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugSettingsDataStore.kt
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugSettingsDataStore.kt:5]: 💡 This could be in the `debugdrawer.data` package since it's related to that feature and other packages don't need to access it.

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:287]: +1 here

File: fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugSettingsDataStore.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugSettingsDataStore.kt:5]: made this change over in https://github.com/mozilla-mobile/firefox-android/pull/4520

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:287]: I'll look into this. I didn't necessarily want to do a debug settings data store observation at this level, but maybe there's another way to do this. Some initial thoughts could be:
- Have a separate observable boolean `fenixOverlayEnabled` that is driven by a series of other conditions (e.g. `debugDrawerEnabled`)
- Scrap the initial idea of having a universal, global Composable for the entire app and just scope it to the debugging. So make `FenixOverlay` into `DebugOverlay` and then the scoped observation would make more sense.

File: fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:51]: so like center along the left edge of the screen? or biased like 50 just left of center?

Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/FenixOverlay.kt:51]: left edge would be my suggestion. i'm mostly just worried that repositioning it might be hard and so punted on, and center screen seems like important real-estate

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:287]: I do think we can probably land something and iterate on it as needed, but the way I'm thinking about it is this:

What if we wanted to use a feature flag or build flag to control whether the overlay was present? For example only including it in debug builds to start. My assumption would be that that would be controlled before the feature package/fragment/whatever got entered, whereas here we are entering the feature first and then determining if we should do anything with it

**Comment by MozillaNoah** [comment]
Recent push includes a rebase from main with the now-landed secret setting, the FAB lowercase fix, and the change to make the fab align to the left edge of the screen.

**Comment by MozillaNoah** [comment]
recent push removes `FenixOverlay`, consolidates the UI to `DebugOverlay` and moves the observation to `HomeActivity`

**Review by MatthewTighe — APPROVED** [review]
LGTM! Let's just monitor perf for the rest of the nightly cycle

**Pull request merged**
